spergelpressmod.f90

```diff
@@ -0,0 +1,255 @@
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Spergel-Press WIMP heat transport module !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+! Contains the functions used in the Spergel Press section of transgen.f90. These are:
+!	-nx_func: Calculates the WIMP density profile using the fgoth interpolation
+! 	-Etrans_sp: calculates the WIMP transported energy (eps_x) given the WIMP temperature (Tx)
+!	-Tx_integral: to be used in newtons_meth
+!	-newtons_meth: solves Tx_integral=0 which defines Tx 
+
+! All units are cgs except tab_r and tab_dr
+! I apologize for the long function calls.
+
+module spergelpressmod
+use capmod
+implicit none
+   
+double precision, parameter :: kB=1.38064852d-16, mnucg=1.6726219e-24
+
+contains
+
+
+function nx_func(T_x, dTdr, mfp, sigma_N, alpha, Nwimps, niso)
+implicit none
+integer, intent(in) :: niso
+double precision, intent(in) :: T_x, Nwimps
+double precision, intent(in) :: dTdr(nlines), mfp(nlines)
+double precision, intent(in) :: alpha(niso), sigma_N(niso)
+double precision :: n_0, fgoth, Tc, K, rchi, integrand, mxg, alphaofR(nlines)
+double precision :: R(nlines), phi(nlines), dR(nlines), dphidr(nlines), n_nuc(niso,nlines)
+double precision :: nx_LTE(nlines), nx_iso(nlines), nx_func(nlines), cumint(nlines), cumNx(nlines)
+integer :: i, j
+! Calculates the isothermal wimp number density
+! I haven't succesfully integrated this function into the code
+! For some reason, it gave problems with Newton's method when I tried to do that (probably a bug)
+
+Tc = tab_T(1) ! K
+R = tab_r*Rsun ! cm
+dR = tab_dr*Rsun ! cm
+phi = -tab_vesc**2/2.d0 ! erg/g
+dphidr = -tab_g ! erg/g/cm
+mxg = mdm*1.782662d-24
+
+! WIMP number density in isothermal approximation
+nx_iso = exp(-mxg*phi/kB/T_x) 
+n_0 = Nwimps/trapz(r, 4.d0*pi*r**2.d0*nx_iso, nlines) ! Normalize so that integral(nx) = Nwimps
+nx_iso = n_0*nx_iso
+
+do i=1,niso
+	n_nuc(i,:) = tab_mfr(:,i)*tab_starrho/AtomicNumber(i)/mnucg
+enddo
+
+rchi = (3.*(kB*Tc)/(2.*pi*GNewt*tab_starrho(1)*mxg))**.5 ! cm
+K = mfp(1)/rchi
+
+cumint(1) = 0.d0
+do i = 1,nlines
+
+! 1) get alpha average
+  alphaofR(i) = sum(alpha*2.d0*sigma_N*n_nuc(:,i))/sum(2.d0*sigma_N*n_nuc(:,i))
+  
+  !perform the integral inside the nx integral
+  integrand = (kB*alphaofR(i)*dTdr(i) + mxg*dphidr(i))/(kB*tab_T(i))
+
+  if (i > 1) then
+  cumint(i) = cumint(i-1) + integrand*dR(i)
+  end if
+
+  nx_LTE(i) = (tab_T(i)/Tc)**(3./2.)*exp(-cumint(i))
+
+  cumNx = cumNx + 4.*pi*R(i)**2*dR(i)*nx_LTE(i)
+
+end do
+
+nx_LTE = nx_LTE/cumNx*nwimps !normalize density
+fgoth = 1./(1.+(K/.4)**2)
+
```

Comment:
nx_func has been changed to nx_isothermal, which just calculates the isothermal WIMP number density in the SP scheme (no fgoth interpolation). I realized that this function cannot be used as nxIso in the GR scheme, because it gives a numerical instability. So the GR scheme just uses the rchi formula for nxIso now.